ccl/multiregionccl: TestMultiRegionDataDriven failed

ccl/multiregionccl.TestMultiRegionDataDriven [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11430622?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11430622?buildTab=artifacts#/) on release-22.2 @ [42fa1fad9af04139575e4f374a8b3f2d4ca453b1](https://github.com/cockroachdb/cockroach/commits/42fa1fad9af04139575e4f374a8b3f2d4ca453b1):


```
             4.542ms      0.054ms    event:sql/conn_executor_exec.go:694 [n5,client=127.0.0.1:42250,user=root] executing after 0 retries, last retry reason: <nil>
             4.591ms      0.049ms    event:sql/conn_executor_exec.go:694 [n5,client=127.0.0.1:42250,user=root] execution starts: distributed engine
             4.654ms      0.063ms    event:sql/distsql_running.go:1749 [n5,client=127.0.0.1:42250,user=root] creating DistSQL plan with isLocal=true
             4.706ms      0.052ms    event:sql/distsql_running.go:664 [n5,client=127.0.0.1:42250,user=root] running DistSQL plan
             4.752ms      0.045ms        === operation:flow _verbose:1 node:5 client:127.0.0.1:42250 user:root
             4.938ms      0.186ms        event:sql/colflow/vectorized_flow.go:241 [n5,client=127.0.0.1:42250,user=root] setting up vectorized flow
             5.067ms      0.129ms        event:sql/colflow/vectorized_flow.go:293 [n5,client=127.0.0.1:42250,user=root] vectorized flow setup succeeded
             5.115ms      0.047ms        event:sql/flowinfra/flow.go:387 [n5,client=127.0.0.1:42250,user=root] starting (0 processors, 0 startables) asynchronously
             5.153ms      0.038ms        event:sql/colflow/vectorized_flow.go:319 [n5,client=127.0.0.1:42250,user=root] running the batch flow coordinator in the flow's goroutine
            18.305ms     13.152ms        event:component:<flow_id:<8418bd41-c979-4839-88b0-9d9be4915b89> type:FLOW id:0 sql_instance_id:5 > net_rx:<latency:<> wait_time:<> deserialization_time:<> tuples_received:<> bytes_received:<> messages_received:<> > net_tx:<tuples_sent:<> bytes_sent:<> messages_sent:<> > kv:<bytes_read:<> tuples_read:<> batch_requests_issued:<> kv_time:<> contention_time:<> num_interface_steps:<> num_internal_steps:<> num_interface_seeks:<> num_internal_seeks:<> num_gets:<> num_scans:<> num_reverse_scans:<> used_streamer:false > exec:<exec_time:<> max_allocated_mem:<> max_allocated_disk:<> consumed_r_u:<> > output:<num_batches:<> num_tuples:<> > flow_stats:<max_mem_usage:<value_plus_one:20481 > max_disk_usage:<value_plus_one:1 > consumed_r_u:<> >
            18.188ms     -0.117ms        structured:{"@type":"type.googleapis.com/cockroach.sql.distsqlrun.ComponentStats","component":{"flowId":"8418bd41-c979-4839-88b0-9d9be4915b89","type":"FLOW","id":0,"sqlInstanceId":5},"netRx":{"latency":{"valuePlusOne":"0s"},"waitTime":{"valuePlusOne":"0s"},"deserializationTime":{"valuePlusOne":"0s"},"tuplesReceived":{"valuePlusOne":"0"},"bytesReceived":{"valuePlusOne":"0"},"messagesReceived":{"valuePlusOne":"0"}},"netTx":{"tuplesSent":{"valuePlusOne":"0"},"bytesSent":{"valuePlusOne":"0"},"messagesSent":{"valuePlusOne":"0"}},"kv":{"bytesRead":{"valuePlusOne":"0"},"tuplesRead":{"valuePlusOne":"0"},"batchRequestsIssued":{"valuePlusOne":"0"},"kvTime":{"valuePlusOne":"0s"},"contentionTime":{"valuePlusOne":"0s"},"numInterfaceSteps":{"valuePlusOne":"0"},"numInternalSteps":{"valuePlusOne":"0"},"numInterfaceSeeks":{"valuePlusOne":"0"},"numInternalSeeks":{"valuePlusOne":"0"},"numGets":{"valuePlusOne":"0"},"numScans":{"valuePlusOne":"0"},"numReverseScans":{"valuePlusOne":"0"},"usedStreamer":false},"exec":{"execTime":{"valuePlusOne":"0s"},"maxAllocatedMem":{"valuePlusOne":"0"},"maxAllocatedDisk":{"valuePlusOne":"0"},"consumedRU":{"valuePlusOne":"0"}},"output":{"numBatches":{"valuePlusOne":"0"},"numTuples":{"valuePlusOne":"0"}},"inputs":[],"flowStats":{"maxMemUsage":{"valuePlusOne":"20481"},"maxDiskUsage":{"valuePlusOne":"1"},"consumedRU":{"valuePlusOne":"0"}}}
             5.167ms    -13.021ms        === operation:batch flow coordinator _unfinished:1 _verbose:1 node:5 client:127.0.0.1:42250 user:root cockroach.flowid:8418bd41-c979-4839-88b0-9d9be4915b89 cockroach.processorid:0
            16.627ms     11.461ms        event:component:<flow_id:<8418bd41-c979-4839-88b0-9d9be4915b89> type:PROCESSOR id:0 sql_instance_id:5 > net_rx:<latency:<> wait_time:<> deserialization_time:<> tuples_received:<> bytes_received:<> messages_received:<> > net_tx:<tuples_sent:<> bytes_sent:<> messages_sent:<> > kv:<bytes_read:<value_plus_one:12 > tuples_read:<value_plus_one:2 > batch_requests_issued:<value_plus_one:2 > kv_time:<value_plus_one:<nanos:6660468 > > contention_time:<value_plus_one:<nanos:1 > > num_interface_steps:<value_plus_one:1 > num_internal_steps:<value_plus_one:1 > num_interface_seeks:<value_plus_one:3 > num_internal_seeks:<value_plus_one:3 > num_gets:<value_plus_one:2 > num_scans:<value_plus_one:1 > num_reverse_scans:<value_plus_one:1 > used_streamer:false > exec:<exec_time:<> max_allocated_mem:<value_plus_one:20481 > max_allocated_disk:<> consumed_r_u:<value_plus_one:1 > > output:<num_batches:<value_plus_one:2 > num_tuples:<value_plus_one:2 > > flow_stats:<max_mem_usage:<> max_disk_usage:<> consumed_r_u:<> >
            16.901ms      0.274ms        event:component:<flow_id:<00000000-0000-0000-0000-000000000000> type:UNSET id:0 sql_instance_id:0 > net_rx:<latency:<> wait_time:<> deserialization_time:<> tuples_received:<> bytes_received:<> messages_received:<> > net_tx:<tuples_sent:<> bytes_sent:<> messages_sent:<> > kv:<bytes_read:<> tuples_read:<> batch_requests_issued:<> kv_time:<> contention_time:<> num_interface_steps:<> num_internal_steps:<> num_interface_seeks:<> num_internal_seeks:<> num_gets:<> num_scans:<> num_reverse_scans:<> used_streamer:false > exec:<exec_time:<> max_allocated_mem:<> max_allocated_disk:<> consumed_r_u:<> > output:<num_batches:<> num_tuples:<> > flow_stats:<max_mem_usage:<> max_disk_usage:<> consumed_r_u:<> >
            16.480ms     -0.421ms        structured:{"@type":"type.googleapis.com/cockroach.sql.distsqlrun.ComponentStats","component":{"flowId":"8418bd41-c979-4839-88b0-9d9be4915b89","type":"PROCESSOR","id":0,"sqlInstanceId":5},"netRx":{"latency":{"valuePlusOne":"0s"},"waitTime":{"valuePlusOne":"0s"},"deserializationTime":{"valuePlusOne":"0s"},"tuplesReceived":{"valuePlusOne":"0"},"bytesReceived":{"valuePlusOne":"0"},"messagesReceived":{"valuePlusOne":"0"}},"netTx":{"tuplesSent":{"valuePlusOne":"0"},"bytesSent":{"valuePlusOne":"0"},"messagesSent":{"valuePlusOne":"0"}},"kv":{"bytesRead":{"valuePlusOne":"12"},"tuplesRead":{"valuePlusOne":"2"},"batchRequestsIssued":{"valuePlusOne":"2"},"kvTime":{"valuePlusOne":"0.006660468s"},"contentionTime":{"valuePlusOne":"0.000000001s"},"numInterfaceSteps":{"valuePlusOne":"1"},"numInternalSteps":{"valuePlusOne":"1"},"numInterfaceSeeks":{"valuePlusOne":"3"},"numInternalSeeks":{"valuePlusOne":"3"},"numGets":{"valuePlusOne":"2"},"numScans":{"valuePlusOne":"1"},"numReverseScans":{"valuePlusOne":"1"},"usedStreamer":false},"exec":{"execTime":{"valuePlusOne":"0s"},"maxAllocatedMem":{"valuePlusOne":"20481"},"maxAllocatedDisk":{"valuePlusOne":"0"},"consumedRU":{"valuePlusOne":"1"}},"output":{"numBatches":{"valuePlusOne":"2"},"numTuples":{"valuePlusOne":"2"}},"inputs":[],"flowStats":{"maxMemUsage":{"valuePlusOne":"0"},"maxDiskUsage":{"valuePlusOne":"0"},"consumedRU":{"valuePlusOne":"0"}}}
            16.647ms      0.167ms        structured:{"@type":"type.googleapis.com/cockroach.sql.distsqlrun.ComponentStats","component":{"flowId":"00000000-0000-0000-0000-000000000000","type":"UNSET","id":0,"sqlInstanceId":0},"netRx":{"latency":{"valuePlusOne":"0s"},"waitTime":{"valuePlusOne":"0s"},"deserializationTime":{"valuePlusOne":"0s"},"tuplesReceived":{"valuePlusOne":"0"},"bytesReceived":{"valuePlusOne":"0"},"messagesReceived":{"valuePlusOne":"0"}},"netTx":{"tuplesSent":{"valuePlusOne":"0"},"bytesSent":{"valuePlusOne":"0"},"messagesSent":{"valuePlusOne":"0"}},"kv":{"bytesRead":{"valuePlusOne":"0"},"tuplesRead":{"valuePlusOne":"0"},"batchRequestsIssued":{"valuePlusOne":"0"},"kvTime":{"valuePlusOne":"0s"},"contentionTime":{"valuePlusOne":"0s"},"numInterfaceSteps":{"valuePlusOne":"0"},"numInternalSteps":{"valuePlusOne":"0"},"numInterfaceSeeks":{"valuePlusOne":"0"},"numInternalSeeks":{"valuePlusOne":"0"},"numGets":{"valuePlusOne":"0"},"numScans":{"valuePlusOne":"0"},"numReverseScans":{"valuePlusOne":"0"},"usedStreamer":false},"exec":{"execTime":{"valuePlusOne":"0s"},"maxAllocatedMem":{"valuePlusOne":"0"},"maxAllocatedDisk":{"valuePlusOne":"0"},"consumedRU":{"valuePlusOne":"0"}},"output":{"numBatches":{"valuePlusOne":"0"},"numTuples":{"valuePlusOne":"0"}},"inputs":[],"flowStats":{"maxMemUsage":{"valuePlusOne":"0"},"maxDiskUsage":{"valuePlusOne":"0"},"consumedRU":{"valuePlusOne":"0"}}}
             5.249ms    -11.398ms        === operation:colbatchscan _unfinished:1 _verbose:1 node:5 client:127.0.0.1:42250 user:root
             5.249ms      0.000ms        [/cockroach.roachpb.Internal/Batch: {count: 2, duration 6ms}]
             5.249ms      0.000ms        [txn coordinator send: {count: 1, duration 6ms}]
             5.249ms      0.000ms        [dist sender send: {count: 1, duration 4ms}]
             5.824ms      0.575ms        event:sql/row/kv_batch_fetcher.go:443 [n5,client=127.0.0.1:42250,user=root] Scan /Table/108/1/1/0
             5.879ms      0.056ms            === operation:txn coordinator send _verbose:1 node:5 client:127.0.0.1:42250 user:root txnID:4a5499bf-42ef-4ef0-aafb-6cb36483b3e7
             5.879ms      0.000ms            [/cockroach.roachpb.Internal/Batch: {count: 2, duration 6ms}]
             5.879ms      0.000ms            [dist sender send: {count: 1, duration 4ms}]
             6.968ms      1.089ms                === operation:dist sender send _verbose:1 node:5 client:127.0.0.1:42250 user:root txn:4a5499bf
             6.968ms      0.000ms                [/cockroach.roachpb.Internal/Batch: {count: 2, duration 6ms}]
             7.103ms      0.135ms                event:kv/kvclient/kvcoord/range_iter.go:183 [n5,client=127.0.0.1:42250,user=root,txn=4a5499bf] querying next range at /Table/108/1/1/0
             7.243ms      0.140ms                event:kv/kvclient/kvcoord/range_iter.go:216 [n5,client=127.0.0.1:42250,user=root,txn=4a5499bf] key: /Table/108/1/1/0, desc: r53:/{Table/108-Max} [(n1,s1):1, (n3,s3):10, (n2,s2):5, (n4,s4):6NON_VOTER, next=13, gen=30]
             7.343ms      0.100ms                event:kv/kvclient/kvcoord/dist_sender.go:1991 [n5,client=127.0.0.1:42250,user=root,txn=4a5499bf] routing to nearest replica; leaseholder not required
             7.616ms      0.273ms                event:kv/kvclient/kvcoord/dist_sender.go:2049 [n5,client=127.0.0.1:42250,user=root,txn=4a5499bf] r53: sending batch 1 Get to (n1,s1):1
             7.674ms      0.058ms                event:rpc/nodedialer/nodedialer.go:163 [n5,client=127.0.0.1:42250,user=root,txn=4a5499bf] sending request to 127.0.0.1:33179
             7.705ms      0.031ms                    === operation:/cockroach.roachpb.Internal/Batch _verbose:1 node:5 client:127.0.0.1:42250 user:root txn:4a5499bf span.kind:client node:5
             8.119ms      0.414ms                    === operation:/cockroach.roachpb.Internal/Batch _verbose:1 node:1 span.kind:server request:Get [/Table/108/1/1/0,/Min), [txn: 4a5499bf], [can-forward-ts]
             8.204ms      0.085ms                    event:server/node.go:1090 [n1] node received request: 1 Get
             9.202ms      0.998ms                    event:kv/kvserver/store_send.go:166 [n1,s1] executing Get [/Table/108/1/1/0,/Min), [txn: 4a5499bf], [can-forward-ts]
             9.285ms      0.083ms                    event:kv/kvserver/replica_send.go:174 [n1,s1,r53/1:/{Table/108-Max}] read-only path
             9.366ms      0.081ms                    event:kv/kvserver/concurrency/concurrency_manager.go:207 [n1,s1,r53/1:/{Table/108-Max}] sequencing request
             9.402ms      0.036ms                    event:kv/kvserver/concurrency/concurrency_manager.go:288 [n1,s1,r53/1:/{Table/108-Max}] acquiring latches
             9.455ms      0.053ms                    event:kv/kvserver/concurrency/concurrency_manager.go:332 [n1,s1,r53/1:/{Table/108-Max}] scanning lock table for conflicting locks
             9.729ms      0.274ms                    event:kv/kvserver/replica_read.go:269 [n1,s1,r53/1:/{Table/108-Max}] executing read-only batch
            10.174ms      0.445ms                    event:kv/kvserver/replica_evaluate.go:544 [n1,s1,r53/1:/{Table/108-Max}] evaluated Get command header:<key:"\364\211\211\210" > , txn="sql txn" meta={id=4a5499bf key=/Min pri=0.00781370 epo=0 ts=1692687432.383366037,0 min=1692687432.383366037,0 seq=0} lock=false stat=PENDING rts=1692687432.383366037,0 wto=false gul=1692687432.883366037,0 : resp=header:<num_keys:1 num_bytes:7 > value:<raw_bytes:"\375\311[\010\n#\002" timestamp:<wall_time:1692687412442292464 > > , err=<nil>
            10.270ms      0.096ms                    event:scan stats: stepped 0 times (0 internal); seeked 2 times (2 internal); block-bytes: (total 15 KiB, cached 15 KiB); points: (count 2 , key-bytes 27 B, value-bytes 87 B, tombstoned: 0 ) ranges: (count 0 ), (contained-points 0 , skipped-points 0 ) evaluated requests: 1 gets, 0 scans, 0 reverse scans
            10.497ms      0.227ms                    event:kv/kvserver/replica_read.go:204 [n1,s1,r53/1:/{Table/108-Max}] read completed
            10.204ms     -0.292ms                    structured:{"@type":"type.googleapis.com/cockroach.roachpb.ScanStats","numInterfaceSeeks":"2","numInternalSeeks":"2","numInterfaceSteps":"0","numInternalSteps":"0","blockBytes":"15451","blockBytesInCache":"15451","keyBytes":"27","valueBytes":"87","pointCount":"2","pointsCoveredByRangeTombstones":"0","rangeKeyCount":"0","rangeKeyContainedPoints":"0","rangeKeySkippedPoints":"0","numGets":"1","numScans":"0","numReverseScans":"0"}
            11.493ms      5.614ms            event:kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:572 [n5,client=127.0.0.1:42250,user=root,txn=4a5499bf] recording span to refresh: /Table/108/1/1/0
            18.563ms     13.857ms    event:sql/conn_executor_exec.go:694 [n5,client=127.0.0.1:42250,user=root] execution ends
            18.603ms      0.040ms    event:sql/conn_executor_exec.go:694 [n5,client=127.0.0.1:42250,user=root] rows affected: 1
            19.911ms      1.308ms        === operation:commit sql txn _verbose:1 node:5 client:127.0.0.1:42250 user:root
            20.209ms      1.606ms    event:sql/conn_executor_exec.go:2191 [n5,client=127.0.0.1:42250,user=root] AutoCommit. err: <nil>
    --- FAIL: TestMultiRegionDataDriven/global_tables (62.73s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #98020 ccl/multiregionccl: TestMultiRegionDataDriven failed [A-kv-distribution A-kv-test-failure-complex C-bug C-test-failure O-robot T-kv branch-master skipped-test]
</p>
</details>
/cc @cockroachdb/multiregion
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMultiRegionDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30829

--- a/pkg/ccl/multiregionccl/datadriven_test.go
+++ b/pkg/ccl/multiregionccl/datadriven_test.go
@@ -113,7 +113,13 @@ func TestMultiRegionDataDriven(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
+	// This test speeds up replication changes by proactively enqueuing replicas
+	// into various queues. This has the benefit of reducing the time taken after
+	// zone config changes, however the downside of added overhead. Disable the
+	// test under deadlock builds, as the test is slow and susceptible to
+	// (legitimate) timing issues on a deadlock build.
 	skip.UnderRace(t, "flaky test")
+	skip.UnderDeadlock(t, "flaky test")
 	skip.WithIssue(t, 98020)
 	ctx := context.Background()
 	datadriven.Walk(t, datapathutils.TestDataPath(t), func(t *testing.T, path string) {

--- a/pkg/ccl/multiregionccl/datadriven_test.go
+++ b/pkg/ccl/multiregionccl/datadriven_test.go
@@ -186,13 +186,17 @@ func TestMultiRegionDataDriven(t *testing.T) {
 				}
 				// Speed up closing of timestamps, in order to sleep less below before
 				// we can use follower_read_timestamp(). follower_read_timestamp() uses
-				// sum of the following settings.
+				// sum of the following settings. Also, disable all kvserver lease
+				// transfers other than those required to satisfy a lease preference.
+				// This prevents the lease shifting around too quickly, which leads to
+				// concurrent replication changes being proposed by prior leaseholders.
 				for _, stmt := range strings.Split(`
 SET CLUSTER SETTING kv.closed_timestamp.target_duration = '0.4s';
 SET CLUSTER SETTING kv.closed_timestamp.side_transport_interval = '0.1s';
 SET CLUSTER SETTING kv.closed_timestamp.propagation_slack = '0.5s';
 SET CLUSTER SETTING kv.allocator.load_based_rebalancing = 'off';
-SET CLUSTER SETTING kv.allocator.load_based_lease_rebalancing.enabled = false
+SET CLUSTER SETTING kv.allocator.load_based_lease_rebalancing.enabled = false;
+SET CLUSTER SETTING kv.allocator.min_lease_transfer_interval = 5m
 `,
 					";") {
 					_, err = sqlConn.Exec(stmt)

